### PR TITLE
improve eslint

### PR DESCRIPTION
**Pull request by septs**
Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/eslint/index.d.ts [hunk 1]
```diff
@@ -431,85 +433,107 @@ export namespace SourceCode {
...
-        YieldExpression?: ((node: ESTree.YieldExpression & NodeParentExtension) => void) | undefined;
-    }
+        meta?: RuleMetaData;
+    }
+
+    type MixedNode = ESTree.Program | NodeTypeMap[keyof NodeTypeMap];
+
+    type NodeTypes = MixedNode["type"];
+
+    interface NodeTypeMap {
```
Comment (septs) [inline types/eslint/index.d.ts:443]: allow user declaration merging add more ast node, e.g: typescript-eslint

Comment (bradzacher) [reply types/eslint/index.d.ts:443]: The problem with this is that it requires the "extension" to redeclare every single node in the AST that uses specific unions.

For any non-trivial extension this means a complete redefinition of the AST.

**Review by bradzacher — CHANGES_REQUESTED** [review]
I don't think this is the correct way to do this.
ESLint doesn't officially support JSX.

Additionally in order support JSX properly, you also need to override the unions used for node props.
For example - `Expression` needs to include `JSXElement`.

This cannot be achieved by just extending the interfaces, and requires updating unmergable unions that live in the `estree` types.

**Comment by septs** [comment]
@bradzacher can I implement replaceable `Node` types through generic?

e.g: `RuleModule<ESTreeJSX.Node>`, `RuleContext<ESTree.Node>`, etc

**Comment by bradzacher** [comment]
You could do that - but you would need to do that for *every single node*.
If you ever forget to pass the generic then your types would be incorrect.
It's a very clunky and error-prone system to implement and manage.
Which is why we at @typescript-eslint just re-implemented the types from the ground up with our own AST.

**Pull request closed without merging**